substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -705,71 +709,82 @@ where
 	async fn submit_events<EventStream>(
 		&mut self,
 		startup_point: &StartupPoint<Block>,
-		mut stream: EventStream,
+		stream: EventStream,
 		sink: Subscription,
 		rx_stop: oneshot::Receiver<()>,
 	) -> Result<(), SubscriptionManagementError>
 	where
-		EventStream: Stream<Item = NotificationType<Block>> + Unpin,
+		EventStream: Stream<Item = NotificationType<Block>> + Unpin + Send,
 	{
-		let mut stream_item = stream.next();
-
-		// The stop event can be triggered by the chainHead logic when the pinned
-		// block guarantee cannot be hold. Or when the client is disconnected.
-		let connection_closed = sink.closed();
-		tokio::pin!(connection_closed);
-		let mut stop_event = futures_util::future::select(rx_stop, connection_closed);
-
-		while let Either::Left((Some(event), next_stop_event)) =
-			futures_util::future::select(stream_item, stop_event).await
-		{
-			let events = match event {
+		// make the stream abortable
+		let (stream, cancel_handle) = futures::stream::abortable(stream);
+
+		// create a channel to propagate error messages
+		let (tx_send, mut tx_receive) = futures::channel::mpsc::channel(1);
+		let mut events = move |event| {
```

Comment:
nit: Would `handle_events` be more explicit here?

Reply:
done